internal/zinc-compile-core/src/main/scala/sbt/internal/inc/ProblemStringFormats.scala

```diff
@@ -0,0 +1,32 @@
+package sbt
+package internal
+package inc
+
+import xsbti.Problem
+import sbt.util.ShowLines
+import sbt.util.InterfaceUtil.jo2o
+
+/**
+ * The string representation of compiler warnings and error messages,
+ * used by LoggerReporter and the logging system.
+ */
+trait ProblemStringFormats {
```

Comment:
Putting typeclass instances into a trait gives most flexibility, because later down the line I can mix and match different formats and make an object a la carte. Although this is slightly different from the Cake pattern, the process is sometimes colloquially called "baking the cake".